cvmfs/sqlitemem.cc

```diff
@@ -0,0 +1,508 @@
+  int32_t usable_size = kArenaSize -
+    (sizeof(void *) + sizeof(AvailBlockCtl) + 1 + sizeof(int32_t));
+  *reinterpret_cast<MallocArena **>(arena_) = this;
+  AvailBlockCtl *free_block =
+    new (arena_ + sizeof(void *) + sizeof(AvailBlockCtl) + 1) AvailBlockCtl();
+  free_block->size = usable_size;
+  free_block->link_next = free_block->link_prev =
+    head_avail_->ConvertToLink(arena_);
+
+  head_avail_->size = 0;
+  head_avail_->link_next = head_avail_->link_prev =
+    free_block->ConvertToLink(arena_);
+
+  // Prevent succeeding blocks from merging
+  *(reinterpret_cast<char *>(free_block) - 1) = kTagReserved;
+  // Final tag: reserved block marker
+  *reinterpret_cast<int32_t *>(arena_ + kArenaSize - sizeof(int32_t)) = -1;
+}
+
+
+MemoryManager::MallocArena::~MallocArena() {
+  sxunmap(arena_, kArenaSize);
+}
+
+
+//------------------------------------------------------------------------------
+
+
+void *MemoryManager::LookasideBufferArena::GetBuffer() {
+  for (unsigned i = 0; i < kNoBitmaps; ++i) {
+    int bit_set = ffs(freemap_[i]);
+    if (bit_set != 0) {
+      freemap_[i] &= ~(1 << (bit_set - 1));  // set bit to zero
+      const int nbuffer = i * sizeof(int) * 8 + bit_set - 1;
+      return reinterpret_cast<char *>(arena_) + nbuffer * kBufferSize;
+    }
+  }
+  return NULL;
+}
+
+
+bool MemoryManager::LookasideBufferArena::IsEmpty() {
+  for (unsigned i = 0; i < kNoBitmaps; ++i) {
+    if (~freemap_[i] != 0)
+      return false;
+  }
+  return true;
+}
+
+
+bool MemoryManager::LookasideBufferArena::Contains(void *buffer) {
+  if ((buffer == NULL) || (buffer < arena_))
+    return false;
+  return (reinterpret_cast<char *>(buffer) - reinterpret_cast<char *>(arena_)) <
+         kArenaSize;
+}
+
+
+MemoryManager::LookasideBufferArena::LookasideBufferArena()
+  : arena_(sxmmap(kArenaSize))
+{
+  // All buffers unused, i.e. all bits set
+  memset(freemap_, 0xFF, kNoBitmaps * sizeof(int));
+}
+
+
+MemoryManager::LookasideBufferArena::~LookasideBufferArena() {
+  sxunmap(arena_, kArenaSize);
+}
+
+
+void MemoryManager::LookasideBufferArena::PutBuffer(void *buffer) {
+  ptrdiff_t nbuffer =
+    (reinterpret_cast<char *>(buffer) - reinterpret_cast<char *>(arena_))
+    / kBufferSize;
+  assert(nbuffer < kBuffersPerArena);
+  const int nfreemap = nbuffer / (sizeof(int) * 8);
+  freemap_[nfreemap] |= 1 << (nbuffer % (sizeof(int) * 8));
+}
+
+
+//------------------------------------------------------------------------------
+
+
+MemoryManager *MemoryManager::instance_ = NULL;
+
+
+/**
+ * Sqlite ensures that size > 0.
+ */
+void *MemoryManager::xMalloc(int size) {
+  return instance_->GetMemory(size);
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL.
+ */
+void MemoryManager::xFree(void *ptr) {
+  instance_->PutMemory(ptr);
+}
+
+
+/**
+ * Sqlite ensures that ptr != NULL and new_size > 0.
+ */
+void *MemoryManager::xRealloc(void *ptr, int new_size) {
+  void *new_ptr = xMalloc(new_size);
+  memcpy(new_ptr, ptr, new_size);
+  xFree(ptr);
```

Comment:
I wonder if it would make sense to implement a `MemoryManager::ReallocMemory()`? If the next memory region is free by chance, this could just extend the allocated region. Questionable how likely this is, though.